api/src/services/files.ts

```diff
@@ -65,39 +71,91 @@ export class FilesService extends ItemsService {
 
 		const fileExtension =
 			path.extname(payload.filename_download!) || (payload.type && '.' + extension(payload.type)) || '';
 
+		// The filename_disk is the FINAL filename on disk
 		payload.filename_disk = primaryKey + (fileExtension || '');
 
+		// Temp filename is used for replacements
+		const tempFilenameDisk = 'temp_' + payload.filename_disk;
+
 		if (!payload.type) {
 			payload.type = 'application/octet-stream';
 		}
 
+		// Used to clean up if something goes wrong
+		const cleanUp = async () => {
+			// If this is a new file upload that failed, we need to delete the DB record
+			if (isReplacement === false) {
+				await this.deleteOne(primaryKey!);
+			}
+
+			try {
+				if (isReplacement === true ){
+					// If this is a replacement that failed, we need to delete the temp file
+					await disk.delete(tempFilenameDisk);
+				} else {
+					// If this is a new file upload that failed, we need to delete the final file
+					await disk.delete(payload.filename_disk!);
+				}
+
+			} catch (err: any) {
+				logger.warn(`Couldn't delete temp file ${tempFilenameDisk}`);
+				logger.warn(err);
+			}
+		};
+
 		try {
-			await storage.location(data.storage).write(payload.filename_disk, stream, payload.type);
+			// If this is a replacement, we'll write the file to a temp location first to ensure we don't overwrite the existing file if something goes wrong
+			if (isReplacement === true) {
+				await disk.write(tempFilenameDisk, stream, payload.type);
+			} else {
+				// If this is a new file upload, we'll write the file to the final location
+				await disk.write(payload.filename_disk, stream, payload.type);
+			}
+
+			// Check if the file was truncated (if the stream ended early) and throw limit error if it was
+			// @ts-expect-error
+			if (stream.truncated === true) {
+				await cleanUp();
+				throw new ContentTooLargeError();
+			}
 		} catch (err: any) {
 			logger.warn(`Couldn't save file ${payload.filename_disk}`);
 			logger.warn(err);
 
-			await this.deleteOne(primaryKey);
+			await cleanUp();
 
-			throw new ServiceUnavailableError({ service: 'files', reason: `Couldn't save file ${payload.filename_disk}` });
+			if (err instanceof ContentTooLargeError) {
+				throw err;
+			} else {
+				throw new ServiceUnavailableError({ service: 'files', reason: `Couldn't save file ${payload.filename_disk}` });
+			}
+		}
+
+		// If the file is a replacement, we need to update the DB record with the new payload, delete the old files, and upgrade the temp file
+		if (isReplacement === true) {
+			await this.updateOne(primaryKey, payload, { emitEvents: false });
+
+			// delete the previously saved file and thumbnails to ensure they're generated fresh
+			for await (const filepath of disk.list(String(primaryKey))) {
+				await disk.delete(filepath);
+			}
+
+			// Upgrade the temp file to the final filename
+			await disk.copy(tempFilenameDisk, payload.filename_disk);
+
+			// Delete the temp file
+			await disk.delete(tempFilenameDisk);
```

Comment:
Instead of a copy and delete should we do a move here? Seems more efficient.

Reply:
Most of the move functions use copy and then delete, but it may be more efficient for some drivers. I made the change.